systems/notification/mailer/pkg/server/mailer.go

```diff
@@ -22,116 +22,124 @@ import (
-		Values:  make(map[string]interface{}),
+	payload := &EmailPayload{
+		To:           req.To,
+		TemplateName: req.TemplateName,
+		Values:       values,
 	}
 
-	for key, value := range values {
-		emailData.Values[key] = value
-	}
-	emailData.Values["EmailID"] = mailId.String()
-	tmpl, err := template.New("email").Parse(emailData.Body)
+	body, err := s.prepareMsg(payload)
 	if err != nil {
-		log.Errorf("Failed to parse email template: %v", err)
 		return nil, err
 	}
 
-	var bodyBuffer bytes.Buffer
-	err = tmpl.Execute(&bodyBuffer, emailData.Values)
+	
+	c, err := smtp.Dial(fmt.Sprintf("%s:%d", s.mailer.Host, s.mailer.Port))
 	if err != nil {
-		log.Errorf("Failed to render email template: %v", err)
 		return nil, err
 	}
+	defer c.Close()
+	config := &tls.Config{
+		ServerName: s.mailer.Host,
+	}
+	if err = c.StartTLS(config); err != nil {
+		return nil, err
+	}
+
+	auth := smtp.PlainAuth("", s.mailer.Username, s.mailer.Password, s.mailer.Host)
 
-	var recipientList []string
-	if len(to) > 0 {
-		recipientList = make([]string, len(to))
-		copy(recipientList, to)
+	// Authenticate with the SMTP server
+	if err = c.Auth(auth); err != nil {
+		return nil, err
 	}
 
-	msg := "From: " + from + "\r\n" +
-		"To: " + strings.Join(recipientList, ",") + "\r\n" +
-		"Subject: " + subject + "\r\n" +
-		"MIME-Version: 1.0\r\n" +
-		"Content-Type: text/html; charset=utf-8\r\n" +
-		"\r\n" +
-		bodyBuffer.String()
+	// Set the sender email address
+	if err = c.Mail(s.mailer.From); err != nil {
+		return nil, err
+	}
 
-	auth := smtp.PlainAuth("", s.mailer.Username, s.mailer.Password, s.mailer.Host)
-	port := strconv.Itoa(s.mailer.Port)
-	err = smtp.SendMail(s.mailer.Host+":"+port, auth, from, recipientList, []byte(msg))
+	// Add the recipient email addresses
+	for _, recipient := range req.To {
+		if err = c.Rcpt(recipient); err != nil {
+			return nil, err
+		}
+	}
+
+	w, err := c.Data()
 	if err != nil {
-		log.Errorf("Failed to send email: %v", err.Error())
-		err = s.mailerRepoRepo.SendEmail(&db.Mailing{
-			MailId:  mailId,
-			Email:   recipientList[0], // Use the first email if only one is provided
-			Subject: subject,
-			Body:    bodyBuffer.String(),
-			SentAt:  sentAt,
-			Status:  "failed",
-		})
 		return nil, err
 	}
+	defer w.Close()
 
-	log.Infof("Email sent successfully to %v", recipientList)
-
-	for _, recipient := range recipientList {
-		err = s.mailerRepoRepo.SendEmail(&db.Mailing{
-			MailId:  mailId,
-			Email:   recipient,
-			Subject: subject,
-			Body:    bodyBuffer.String(),
-			SentAt:  sentAt,
-			Status:  "sent",
-		})
-		if err != nil {
-			log.Error("Error while sending email" + err.Error())
-			return nil, grpc.SqlErrorToGrpc(err, "Failed to send email")
-		}
+	_, err = w.Write(body.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.Quit()
+	if err != nil {
+		log.Errorln("Email not sent ", err.Error())
 	}
 
-	response := &pb.SendEmailResponse{
-		Message: "Email sent successfully",
+	mailId := uuid.NewV4()
```

Comment:
Also, where do we store emails inside the DB?

Reply:
Fixed